operators/src/processing/temporal_raster_aggregation.rs

```diff
@@ -145,119 +158,71 @@ where
-
-                let s = RasterOverlapAdapter::<'a, P, _, _>::new(
-                    &self.source,
-                    query,
-                    self.tiling_specification,
-                    ctx,
-                    spec,
-                );
-
-                Ok(s.boxed())
-            }
+            } => Ok(self
+                .create_subquery(fold_future::<P, MaxAccFunction>, P::min_value())
+                .into_raster_overlap_adapter(&self.source, query, ctx, self.tiling_specification)
+                .boxed()),
             AggregationType::FirstValid => {
-                let spec = TemporalRasterAggregationSubQuery {
-                    fold_fn: no_data_ignoring_fold_future::<P, FirstValidAccFunction>,
-                    no_data_value: self.no_data_value,
-                    initial_value: self.no_data_value.unwrap_or_else(P::zero), // TODO: what to do if there is no nodata?
-                    step: self.window,
-                };
-
-                let s = RasterOverlapAdapter::<'a, P, _, _>::new(
-                    &self.source,
-                    query,
-                    self.tiling_specification,
-                    ctx,
-                    spec,
-                );
-
-                Ok(s.boxed())
+                let no_data_value =
+                    self.no_data_value
+                        .ok_or(error::Error::OperationRequiresNoDataValue {
+                            operation_name: "TemporalRasterAggregation::FirstValid".to_string(),
+                        })?;
+                Ok(self
+                    .create_subquery(
+                        no_data_ignoring_fold_future::<P, FirstValidAccFunction>,
+                        no_data_value,
+                    )
+                    .into_raster_overlap_adapter(
+                        &self.source,
+                        query,
+                        ctx,
+                        self.tiling_specification,
+                    )
+                    .boxed())
             }
             AggregationType::LastValid => {
-                let spec = TemporalRasterAggregationSubQuery {
-                    fold_fn: no_data_ignoring_fold_future::<P, LastValidAccFunction>,
-                    no_data_value: self.no_data_value,
-                    initial_value: self.no_data_value.unwrap_or_else(P::zero), // TODO: what to do if there is no nodata?
-                    step: self.window,
-                };
-
-                let s = RasterOverlapAdapter::<'a, P, _, _>::new(
-                    &self.source,
-                    query,
-                    self.tiling_specification,
-                    ctx,
-                    spec,
-                );
-
-                Ok(s.boxed())
+                let no_data_value =
+                    self.no_data_value
+                        .ok_or(error::Error::OperationRequiresNoDataValue {
+                            operation_name: "TemporalRasterAggregation::FirstValid".to_string(),
```

Comment:
wrong operation_name. Maybe it would make more sense to have a nested error here. Maybe a TemporalRasterAggreation error with a nested  MissingNoData error?